lib/Frontend/Frontend.cpp

```diff
@@ -261,236 +264,305 @@ ModuleDecl *CompilerInstance::getMainModule() {
-                                               Invocation.getSourceFileKind(),
-                                               MainBufferID, modImpKind,
-                                               KeepTokens);
-    MainModule->addFile(*MainFile);
-    addAdditionalInitialImports(MainFile);
+ModuleDecl *CompilerInstance::importUnderlyingModule() {
+  SharedTimer timer("performSema-importUnderlyingModule");
+  ModuleDecl *objCModuleUnderlyingMixedFramework =
+      static_cast<ClangImporter *>(Context->getClangModuleLoader())
+          ->loadModule(SourceLoc(),
+                       std::make_pair(MainModule->getName(), SourceLoc()));
+  if (objCModuleUnderlyingMixedFramework)
+    return objCModuleUnderlyingMixedFramework;
+  Diagnostics.diagnose(SourceLoc(), diag::error_underlying_module_not_found,
+                       MainModule->getName());
+  return nullptr;
+}
 
-    if (MainBufferID == PrimaryBufferID)
-      setPrimarySourceFile(MainFile);
+ModuleDecl *CompilerInstance::importBridgingHeader() {
+  SharedTimer timer("performSema-importBridgingHeader");
+  const StringRef implicitHeaderPath =
+      Invocation.getFrontendOptions().ImplicitObjCHeaderPath;
+  auto clangImporter =
+      static_cast<ClangImporter *>(Context->getClangModuleLoader());
+  if (implicitHeaderPath.empty() ||
+      clangImporter->importBridgingHeader(implicitHeaderPath, MainModule))
+    return nullptr;
+  ModuleDecl *importedHeaderModule = clangImporter->getImportedHeaderModule();
+  assert(importedHeaderModule);
+  return importedHeaderModule;
+}
+
+void CompilerInstance::getImplicitlyImportedModules(
+    SmallVectorImpl<ModuleDecl *> &importModules) {
+  SharedTimer timer("performSema-getImplicitlyImportedModules");
+  for (auto &ImplicitImportModuleName :
+       Invocation.getFrontendOptions().ImplicitImportModuleNames) {
+    if (Lexer::isIdentifier(ImplicitImportModuleName)) {
+      auto moduleID = Context->getIdentifier(ImplicitImportModuleName);
+      ModuleDecl *importModule =
+          Context->getModule(std::make_pair(moduleID, SourceLoc()));
+      if (importModule) {
+        importModules.push_back(importModule);
+      } else {
+        Diagnostics.diagnose(SourceLoc(), diag::sema_no_import,
+                             ImplicitImportModuleName);
+        if (Invocation.getSearchPathOptions().SDKPath.empty() &&
+            llvm::Triple(llvm::sys::getProcessTriple()).isMacOSX()) {
+          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk);
+          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk_xcrun);
+        }
+      }
+    } else {
+      Diagnostics.diagnose(SourceLoc(), diag::error_bad_module_name,
+                           ImplicitImportModuleName, false);
+    }
   }
+}
 
-  bool hadLoadError = false;
+void CompilerInstance::createREPLFile(
+    const ImplicitImports &implicitImports) const {
+  SharedTimer timer("performSema-createREPLFile");
```

Comment:
Since you asked me to look at timers, this one isn't particularly interesting, since it's not calling anything that does unbounded work.

Reply:
Deleted.